Code/Tools/FBuild/FBuildCore/Helpers/CompilationDatabase.cpp

```diff
@@ -0,0 +1,215 @@
+#include "Tools/FBuild/FBuildCore/PrecompiledHeader.h"
+
+#include "CompilationDatabase.h"
+
+#include "Tools/FBuild/FBuildCore/FBuild.h"
+#include "Tools/FBuild/FBuildCore/Graph/CompilerNode.h"
+#include "Tools/FBuild/FBuildCore/Graph/DirectoryListNode.h"
+#include "Tools/FBuild/FBuildCore/Graph/LibraryNode.h"
+#include "Tools/FBuild/FBuildCore/Graph/NodeGraph.h"
+#include "Tools/FBuild/FBuildCore/Graph/ObjectListNode.h"
+#include "Tools/FBuild/FBuildCore/Graph/ObjectNode.h"
+#include "Tools/FBuild/FBuildCore/Graph/UnityNode.h"
+
+// Core
+#include "Core/Strings/AStackString.h"
+
+// CONSTRUCTOR
+//------------------------------------------------------------------------------
+CompilationDatabase::CompilationDatabase()
+: m_Output( 4 * 1024 * 1024 )
+{
+    m_DirectoryEscaped = FBuild::Get().GetWorkingDir();
+    m_DirectoryEscaped.JSONEscape();
+}
+
+// DESTRUCTOR
+//------------------------------------------------------------------------------
+CompilationDatabase::~CompilationDatabase() = default;
+
+// Generate
+//------------------------------------------------------------------------------
+const AString & CompilationDatabase::Generate( const NodeGraph & nodeGraph, Dependencies & dependencies )
+{
+    m_Output += "[\n";
+
+    const size_t numNodes = nodeGraph.GetNodeCount();
+    Array< bool > visited( numNodes, false );
+    visited.SetSize( numNodes );
+    memset( visited.Begin(), 0, numNodes );
+
+    VisitNodes( dependencies, visited );
+
+    // Remove last comma
+    for ( uint32_t i = m_Output.GetLength() - 1; i != 0; --i )
+    {
+        const char c = m_Output[ i ];
+        if ( c == ',' )
+        {
+            m_Output[ i ] = '\n';
+            m_Output.SetLength( i + 1 );
+            break;
+        }
+        if ( ( c != '\n' ) && ( c != '\r' ) && ( c != ' ' ) && ( c != '\t' ) )
+        {
+            break;
+        }
+    }
+
+    m_Output += "]\n";
+
+    return m_Output;
+}
+
+// VisitNodes
+//------------------------------------------------------------------------------
+void CompilationDatabase::VisitNodes( const Dependencies & dependencies, Array< bool > & visited )
+{
+    for ( const Dependency & dep : dependencies )
+    {
+        Node * node = dep.GetNode();
+
+        // Skip already visited nodes
+        uint32_t nodeIndex = node->GetIndex();
+        ASSERT( nodeIndex != INVALID_NODE_INDEX );
+        if ( visited[ nodeIndex ] )
+        {
+            continue;
+        }
+        visited[ nodeIndex ] = true;
+
+        VisitNodes( node->GetPreBuildDependencies(), visited );
+        VisitNodes( node->GetStaticDependencies(), visited );
+        VisitNodes( node->GetDynamicDependencies(), visited );
+
+        switch ( node->GetType() )
+        {
+            case Node::DIRECTORY_LIST_NODE:
+            {
+                // Build directory list node to populate its file list
+                node->CastTo< DirectoryListNode >()->DoBuild( nullptr );
```

Comment:
DirectoryListNodes can have PreBuildDependencies, which won't be correctly traversed by this manual invocation of the build.

Reply:
Yes, this is a problem. But I think it is not a severe problem because:
1. Most source files are written manually and not generated, so actual impact will be low.
2. Due to ways in which compilation databases are used (providing includes/defines to static analysis/refactoring tools and IDEs) they are much less useful for generated files.
3. If the file generated by a pre-build dependency doesn't exist then the corresponding entry in the compilation database is useless, but if the file exists then it will be found by the DirectoryNode even without building pre-build dependencies.
4. If somebody really wants files generated by pre-build dependencies to be present in the generated compilation database all he/she has to do is to as FASTBuild to execute that part of the build just once (not the whole build, and not each time `-compdb` is used).

So to me it seems that there is little value in solving this problem.
I think issuing a warning like "Skipping pre-build dependency NAME, files generated by it may or may not be present in the generated compilation database." when we encounter ObjectListNode, LibraryNode or UnityNode with a pre-build dependency will be enough.